### ISSUE TITLE
Registry proxy broken due to `hostPort` not working with CNI

### ISSUE THREAD

**Issue by etotheipi** [kubernetes/kubernetes#44564]
**Kubernetes version** (use `kubectl version`):  1.6.1
- **Cloud provider or hardware configuration**:  Metal/OpenStack
- **OS** (e.g. from /etc/os-release):  Centos:7
- **Kernel** (e.g. `uname -a`):  4.10.9
- **Install tools**:  kargo, weave

**What happened**:
Followed directions to setup in-cluster registry with proxy:
https://github.com/kubernetes/kubernetes/tree/master/cluster/addons/registry
Local daemonset proxy fails to open up port 5000 on hosts when using CNI networking (in this case, weave). 

This is a known issue that is being worked on, due to CNI networking not support `hostPort`. https://github.com/kubernetes/kubernetes/issues/31307

**How to fix it**:
The issue is that the registry proxy is run as a pod on the cluster/CNI network using a `hostPort` setting, expecting to be able to open that port for listening on the host. Unfortunately `hostPort`s are broken (silently) when using any CNI networking plugin.  (The one exception I think is kubenet which has integrated a fix into the plugin itself).

I have a workaround which is kinda messy:

I have created a new registry proxy which uses `hostNetwork: true` (avoids Weave/CNI altogether), and the port can be exposed properly on the host. But then DNS resolution to the registry doesn't work, but its IP *is* routable. I added a unique label to my registry pod, then used the service account credentials to find its IP address and copy that into the nginx proxy config. I am having strange problems when I try to push or pull large images, but for the most part it works.  (perhaps the nginx proxy needs to be tweaked)

Unfortunately, I have this proxy code in a private repo, but I can copy the core components here:

Proxy is deployed as a daemonset.  In the entrypoint script, I use svc account credentials in the pod to search the `default` namespace for a pod with label `theclusterregistry=1`.  Registry pod is given this label.  Here's how I get the pod (replace $API_SERVER with the host:port, which might have to be hardcoded)

```
curl \
   --cacert /var/run/secrets/kubernetes.io/serviceaccount/ca.crt \
   -H "Authorization: Bearer $(cat /var/run/secrets/kubernetes.io/serviceaccount/token)" \
   https://$API_SERVER/api/v1/namespaces/default/pods?labelSelector=theclusterregistry=1 | jq '.items[0].status.podIP
```

This is run in a loop until a non-null IP is returned.  I strip its double-quotes (a jq artifact) and then template fill the %HOST% in the nginx config (`/etc/nginx/conf.d/default.conf.in`):

```
# Docker registry proxy for api version 2
upstream docker-registry {
    server %HOST%:5000;
}

server {
    listen 127.0.0.1:5000;
    server_name localhost;

    # disable any limits to avoid HTTP 413 for large image uploads
    client_max_body_size 0;

    # required to avoid HTTP 411: see Issue #1486 (https://github.com/docker/docker/issues/1486)
    chunked_transfer_encoding on;

    location / {
        proxy_pass                          http://docker-registry;
        proxy_set_header  Host              $http_host;   # required for docker client's sake
        proxy_set_header  X-Real-IP         $remote_addr; # pass on real client's IP
        proxy_set_header  X-Forwarded-For   $proxy_add_x_forwarded_for;
        proxy_set_header  X-Forwarded-Proto $scheme;
        proxy_read_timeout                  900;
    }
}
```

It's messy but it works.  Mostly.  I copied and modified the nginx conf from the original proxy.  It seems to really struggle with big images, and I'm unclear why (it frequently freezes once all the data is uploaded/downloaded but before declaring success).  That may be a separate issue to file.

**Comment by kargakis** [comment]
@kubernetes/sig-network-bugs

**Comment by thockin** [comment]
CNI now has chains of plugins, including one for hostPort.  This should be closed in 1.7
